src/lib/peer_trust/peer_trust.ml

```diff
@@ -31,30 +30,243 @@ module type S = sig
+        | Big_credit -> Trust_response.Trust_increase 0.2
+    end
+
+    module Peer_trust_test = Make (Int) (Mock_now) (Action) (Db)
+
+    (* We want to check the output of the pipe in these tests, but it's
+       synchronous, so we need to read from it in a different "thread",
+       otherwise it would hang. *)
+    let ban_pipe_out = ref []
+
+    let assert_ban_pipe expected =
+      [%test_eq: int list] expected !ban_pipe_out ;
+      ban_pipe_out := []
+
+    let setup_mock_db () =
+      let res = Peer_trust_test.create ~db_dir:"fake" in
+      don't_wait_for
+      @@ Strict_pipe.Reader.iter_without_pushback res.bans_reader ~f:(fun v ->
+             ban_pipe_out := v :: !ban_pipe_out ) ;
+      res
+
+    let nolog = Logger.null ()
+
+    let%test "Peers are unbanned and have 0 trust at initialization" =
+      let db = setup_mock_db () in
+      match Peer_trust_test.lookup db 0 with
+      | {trust= 0.0; banned= Unbanned} -> assert_ban_pipe [] ; true
+      | _ -> false
+
+    let%test "Insta-bans actually do so" =
+      let db = setup_mock_db () in
+      Thread_safe.block_on_async_exn (fun () ->
+          Peer_trust_test.record db nolog 0 Insta_ban ) ;
+      match Peer_trust_test.lookup db 0 with
+      | {trust= -1.0; banned= Banned_until time} ->
+          [%test_eq: Time.t] time
+          @@ Time.add !Mock_now.current_time Time.Span.day ;
+          assert_ban_pipe [0] ;
+          true
+      | _ -> false
+
+    let%test "trust decays by half in 24 hours" =
+      let db = setup_mock_db () in
+      Thread_safe.block_on_async_exn (fun () ->
+          Peer_trust_test.record db nolog 0 Action.Big_credit ) ;
+      match Peer_trust_test.lookup db 0 with
+      | {trust= start_trust; banned= Unbanned} -> (
+          Mock_now.advance Time.Span.day ;
+          assert_ban_pipe [] ;
+          match Peer_trust_test.lookup db 0 with
+          | {trust= decayed_trust; banned= Unbanned} ->
+              (* N.b. the floating point equality operator has a built in
+                 tolerance i.e. it's approximate equality. *)
+              decayed_trust =. start_trust /. 2.0
+          | _ -> false )
+      | _ -> false
+
+    let do_constant_rate rate f =
+      (* Simulate running the function at the specified rate, in actions/sec,
+         for a week. *)
+      let instances = Float.to_int @@ (60. *. 60. *. 24. *. 7. *. rate) in
+      let rec go n =
+        if n < instances then (
+          let%bind () = f () in
+          Mock_now.advance Time.Span.(second / rate) ;
+          go (n + 1) )
+        else Deferred.unit
+      in
+      go 0
+
+    let act_constant_rate db rate act =
+      (* simulate performing an action at the specified rate for a week *)
+      do_constant_rate rate (fun () -> Peer_trust_test.record db nolog 0 act)
+
+    let%test "peers don't get banned for acting at the maximum rate" =
+      let db = setup_mock_db () in
+      Thread_safe.block_on_async_exn (fun () ->
```

Comment:
`block_on_async` tricks the OCaml inliner so it's better if you only do this at the top of the test and stay within the deferred monad throughout.

```ocaml
Thread_safe.block_on_async_exn (fun () ->
  let db = setup_mock_db() in
  let%map () = act_constant_rate db 1. Action.Slow_punish in
  (* ... *)
)
```

Please do this for the rest of the tests too.